libkbfs/backpressure_disk_limiter.go

```diff
@@ -15,52 +15,232 @@ import (
 	"golang.org/x/net/context"
 )
 
-// backpressureDiskLimiter is an implementation of diskLimiter that
-// uses backpressure to slow down block puts before they hit the disk
-// limits.
+// backpressureTracker keeps track of the variables used to calculate
+// backpressure. It keeps track of a generic resource (which can be
+// either bytes or files).
 //
-// Let J be the (approximate) byte usage of the journal and F be the
-// free bytes on disk. Then we want to enforce
+// Let U be the (approximate) resource usage of the journal and F be
+// the free resources. Then we want to enforce
 //
-//   J <= min(k(J+F), L),
+//   U <= min(k(U+F), L),
 //
-// where 0 < k <= 1 is some fraction, and L > 0 is the absolute byte
-// usage limit. But in addition to that, we want to set thresholds 0
-// <= m <= M <= 1 such that we apply proportional backpressure (with a
-// given maximum delay) when
+// where 0 < k <= 1 is some fraction, and L > 0 is the absolute
+// resource usage limit. But in addition to that, we want to set
+// thresholds 0 <= m <= M <= 1 such that we apply proportional
+// backpressure (with a given maximum delay) when
 //
-//   m <= max(J/(k(J+F)), J/L) <= M,
+//   m <= max(U/(k(U+F)), J/L) <= M,
 //
 // which is equivalent to
 //
-//   m <= J/min(k(J+F), L) <= M.
-//
-// TODO: Also do limiting based on file counts.
+//   m <= U/min(k(U+F), L) <= M.
+type backpressureTracker struct {
+	// minThreshold is m in the above.
+	minThreshold float64
+	// maxThreshold is M in the above.
+	maxThreshold float64
+	// limitFrac is k in the above.
+	limitFrac float64
+	// limit is L in the above.
+	limit int64
+
+	// used is U in the above.
+	used int64
+	// free is F in the above.
+	free int64
+
+	semaphoreMax int64
+	semaphore    *kbfssync.Semaphore
+}
+
+func newBackpressureTracker(minThreshold, maxThreshold, limitFrac float64,
+	limit, initialFree int64) (*backpressureTracker, error) {
+	if minThreshold < 0.0 {
+		return nil, errors.Errorf("minThreshold=%f < 0.0",
+			minThreshold)
+	}
+	if maxThreshold < minThreshold {
+		return nil, errors.Errorf(
+			"maxThreshold=%f < minThreshold=%f",
+			maxThreshold, minThreshold)
+	}
+	if 1.0 < maxThreshold {
+		return nil, errors.Errorf("1.0 < maxThreshold=%f",
+			maxThreshold)
+	}
+	if limitFrac < 0.01 {
+		return nil, errors.Errorf("limitFrac=%f < 0.01", limitFrac)
+	}
+	if limitFrac > 1.0 {
+		return nil, errors.Errorf("limitFrac=%f > 1.0", limitFrac)
+	}
+	if limit < 0 {
+		return nil, errors.Errorf("limit=%d < 0", limit)
+	}
+	if initialFree < 0 {
+		return nil, errors.Errorf("initialFree=%d < 0", initialFree)
+	}
+	bt := &backpressureTracker{
+		minThreshold, maxThreshold, limitFrac, limit,
+		0, initialFree, 0, kbfssync.NewSemaphore(),
+	}
+	bt.updateSemaphoreMax()
+	return bt, nil
+}
+
+// max returns the resource limit, taking into account the amount of
+// free resources left. This is min(k(U+F), L).
+func (bt backpressureTracker) max() float64 {
```

Comment:
Confusing to call this `max`, especially when it's taking a min.  Can it be called `currLimit()` maybe?